std/random.d

```diff
@@ -1524,6 +1518,381 @@ version (AnyARC4Random)
+    {
+        ulong[2] s = void;
+        // Try to initialize using system entropy.
+        if (oneTimeGenRandom(&s, typeof(s).sizeof) != typeof(s).sizeof)
+        {
+            // Fall back to tid+pid+time seed.
+            auto x = fallbackSeed();
+            enum goldenGamma = 0x9e37_79b9_7f4a_7c15UL; // odd number closest to 2^^64 / golden ratio
+            // Stafford Mix07 for seed.
+            ulong y = (x += goldenGamma);
+            y = (y ^ (y >>> 30)) * 0x16a6_ac37_883a_f045UL;
+            y = (y ^ (y >>> 26)) * 0xcc9c_31a4_2746_86a5UL;
+            s[0] += (y ^ (y >>> 32));
+            // Stafford Mix06 for gamma.
+            ulong z = (x += goldenGamma);
+            z = (z ^ (z >>> 31)) * 0x69b0_bc90_bd9a_8c49UL;
+            z = (z ^ (z >>> 27)) * 0x3d5e_661a_2a77_868dUL;
+            z = z ^ (z >>> 30);
+            s[1] += z;
+        }
+        outSeed = s[0];
+        immutable ulong z = s[1] | 1UL; // Gamma must be odd for maximum period.
+        // Avoid long runs of 1s or 0s in gamma.
+        import core.bitop : popcnt;
+        immutable ulong alt = z ^ 0xaaaa_aaaa_aaaa_aaaaUL;
+        outGamma = popcnt(z ^ (z >>> 1)) < 24 ? alt : z;
+    }
+
+    import core.atomic : has64BitCAS;
+
+    /+
+    In the absence of a convenient arc4random-like interface use a
+    shared static SplitMix generator. Described by Guy L. Steele Jr.,
+    Doug Lea, and Christine H. Flood in "Fast Splittable Pseudorandom
+    Number Generators" (2014). As suggested in the paper use David
+    Stafford's superior parameterizations of the MurmurHash3 64-bit
+    finalizer from
+    http://zimbry.blogspot.com/2011/09/better-bit-mixing-improving-on.html.
+    +/
+    static if (has64BitCAS)
+    private ulong globalSplitMix()()
+    {
+        import core.atomic : atomicLoad, atomicOp, atomicStore, MemoryOrder;
+        shared static ulong _globalSeed;
+        shared static ulong _globalGamma;
+        ulong gamma = atomicLoad!(MemoryOrder.raw)(_globalGamma);
+        if (!gamma)
+        {
+            synchronized
+            {
+                gamma = atomicLoad(_globalGamma);
+                if (!gamma)
+                {
+                    // gamma hasn't been set yet, so initialize.
+                    ulong initialSeed = void;
+                    oneTimeInitSeedAndGamma(initialSeed, gamma);
+                    atomicStore(_globalSeed, initialSeed);
+                    atomicStore(_globalGamma, gamma);
+                }
+            }
+            assert(gamma);
+        }
+        // Stafford Mix05
+        ulong x = atomicOp!"+="(_globalSeed, gamma);
+        x = (x ^ (x >>> 31)) * 0x79c1_35c1_674b_9addUL;
+        x = (x ^ (x >>> 29)) * 0x54c7_7c86_f691_3e45UL;
+        return x ^ (x >>> 30);
+    }
+    /+
+    Similar to above but without 64-bit atomic ops. (All current platforms
+    supported by DMD or LDC have 64-bit atomic ops, but in the future some
+    might not.)
```

Comment:
This would appear to imply that:

* this entire `!has64BitCAS` implementation is never going to be used

* if phobos is ported to a platform without 64-bit atomics, this code will silently start working despite never having been used in production before

... is this correct?

Reply:
>this entire !has64BitCAS implementation is never going to be used

GDC supports / will support a wider variety of architectures than either DMD or LDC.

>if phobos is ported to a platform without 64-bit atomics, this code will silently start working despite never having been used in production before

On architectures without 64-bit atomics the code will be unit tested as part of `unpredictableSeed`. It may be the first time someone has used it "in production" but it won't be the first time it has been used.

Reply:
> GDC supports / will support a wider variety of architectures than either DMD or LDC.

Yes, but that's not the point.  It's great to think of other architectures' needs ahead of time, but it is not healthy to have code that is never exercised hanging around the repo and silently coming into play at some point in the future without warning.  That's a great recipe for nasty, hard-to-detect bugs.

This is why it's so problematic that the splitmix implementation is not well separated and tested from the indeterminate parts of `unpredictableSeed`.

> On architectures without 64-bit atomics the code will be unit tested as part of `unpredictableSeed`.

(1) the code will never be tested or exercised _until_ then.  This is a problem: we should not accept untested code into the repo.

(2) The unittests that exist simply do not cover the things that need testing, and indeed cannot, because of the black-box nature of `unpredictableSeed`.  It would be trivial for a completely flawed seed-generation mechanism to "pass" those tests.

Reply:
The code will never come into play without warning, it will come into play _after passing unit tests_. This objection could also be trivially addressed by always running this code branch in `version (unittest)` even on architectures where it is not used, but that would be pointless.

>(2) The unittests that exist simply do not cover the things that need testing, and indeed cannot, because of the black-box nature of `unpredictableSeed`.

If you think that then your idea of what needs testing is incorrect. We don't need to test that arithmetic works, and we don't need to test that unpredictableSeed's output is identical to the output of any published PRNG because that is not part of its public contract.

>It would be trivial for a completely flawed seed-generation mechanism to "pass" those tests.

For instance, the current seed-generation mechanism you argue we should keep.

Reply:
> The code will never come into play without warning, it will come into play after passing unit tests.

Which tests, exactly?

> We don't need to test that arithmetic works

We _should_ test that SplitMix does what we expect of it, with comparison to the output of a reference implementation.  Otherwise, how do we guard against the possibility that you made a typo when writing this (currently un-exercised) code?  Or how do we guard against the possibility that someone else introduces a mistake in future?

> > It would be trivial for a completely flawed seed-generation mechanism to "pass" those tests.
>
> For instance, the current seed-generation mechanism you argue we should keep.

In which case, the unittests do not verify the quality of randomness of the `unpredictableSeed` output.  Which means that _the behaviour is not tested_.

Reply:
>In which case, the unittests do not verify the quality of randomness of the `unpredictableSeed` output. Which means that _the behaviour is not tested._

If the current version is equally 'untested' in that regard, then this is not a rational basis for obstructing this PR.

Reply:
> If the current version is equally 'untested' in that regard, then this is not a rational basis for obstructing this PR.

On the contrary: in the absence of validation, it's far preferable to stick with an imperfect implementation whose behaviour is known by experience, than a heavily modified version that is much harder to understand and reason about.

For the record, I'm actually trying some stuff out right now to see if I can separate out `SplitMix` in a way that's reasonably testable, without interfering with your wish that it be a `shared` instance.